### ISSUE TITLE
libccv cifar-10.c

### ISSUE THREAD

**Issue by nirwandogra** [liuliu/ccv#97]
Hey I am training cifar-10.c using the cifar dataset.
I want to inform the maker that I think there is an error in the backpropagate function in the ccv_convnet.c file.
Plz someone help.
Its urgent

**Comment by nirwandogra** [comment]
It is also possible that there is some error in the ccv_convnet_supervised_train this function. The miss rate is not going down so sir plz plz plz help.

**Comment by liuliu** [comment]
Hey, it would be helpful to provide your ./lib/config.mk file, environment, and which cifar-10 data file you use.
The cifar-10.c file is not maintained, I will revive it at some point (as a tutorial material), and the file itself is not even mentioned in the doc :smile: 
Also, fix your CapsLock key on the keyboard would help the readability too.

**Comment by nirwandogra** [comment]
hello ,i am using batch1 file of the cifar-10 data file.........
https://github.com/liuliu/ccv/blob/unstable/bin/cifar-10.c i am running this file of urs to train ..........
Note that it gives assertion fail since covnet is 31_31 and images are 32_32 but i changed the code and constructed the densematrix of only 31*31 size.......it passes the assertion without fail but but 
IT is giving a very high miss rate of upto 89-90% and doesnt goes down..........all parameters remain the same as given in your code ...........i will upload my  ./lib/config.mk in the morning ist plz since now i am operating windows.............kindly help me out on this since i want to train my own dataset and u have a brilliant library.
apart from this sorry for the capitals..............i have mentioned this post also on your site libccv.org....so kindly help me....

**Comment by nirwandogra** [comment]
![20140715_130654](https://cloud.githubusercontent.com/assets/4733368/3589027/f2efa4d8-0c49-11e4-961b-6472e97de724.jpg)
plz see this..ill upload the make file tomorrow.i had taken a screenshot earlier in the morning.
some of the things printed were my debug statements and not present in your code......

**Comment by nirwandogra** [comment]
https://github.com/nirwandogra/HACKERRANK/blob/master/PEGSOLITAIRE.cpp
this is my makefile in the bin directory
plz dont go by the name of the link........i changed the contents of the file to show the  makefile....plz plz check

**Comment by liuliu** [comment]
Please include the generated config.mk file, not the makefile.

**Comment by nirwandogra** [comment]
CC := clang
AR := ar
NVCC := 
CUDA_OBJS := 
DEFINE_MACROS := -D HAVE_SSE2 -D HAVE_LIBPNG -D HAVE_LIBJPEG -D HAVE_CBLAS -D HAVE_GSL
CFLAGS := -msse2 $(DEFINE_MACROS)
NVFLAGS := --use_fast_math -arch=sm_30 $(DEFINE_MACROS)
LDFLAGS := -lm -lpng -ljpeg -lblas -lgsl -lgslcblas

**Comment by nirwandogra** [comment]
the above is my config.mk file

**Comment by nirwandogra** [comment]
Sir plz reply soon.

**Comment by liuliu** [comment]
First thing you should know is that the CPU implementation of training is not maintained as of in current version, and cifar-10.c is not maintained too (it is not mentioned in the doc). That is the reason why you encountered the assertion first (which is good, gives the red-flag of using the file at your own risk).

This stopped to work due to weight relayout cache introduced earlier with SSE2 / NEON implementation, I fixed it in master now.

Now, how to make the CPU version work.

The main difference of the CPU version and the GPU version is that the CPU version doesn't support Dropout (which is irrelevant in this case), and doesn't compute mean activity (which is relevant in this case). The learn rate is a bit different between CPU version and GPU version (now fixed with new push).

Basically, to get sensible data for CPU version, you need to adjust the learn rate and also, subtract the input by 128 (thus, to emulate subtract the mean activity).

See this diff about how this works: https://gist.github.com/liuliu/b38942557cb7050d51e0

Again, this file is not maintained as of now, I do plan to make this file work again, but that is low-pri (after R-CNN and SURF cascade face detector), and please use the CPU implementation at your own risk.
